parl/framework/policy_distribution.py

```diff
@@ -69,30 +89,28 @@ class Deterministic(PolicyDistribution):
     def __init__(self, dist):
         super(Deterministic, self).__init__(dist)
         ## For deterministic action, we only support continuous ones
+        assert isinstance(dist, Variable)
         assert dist.dtype == convert_np_dtype_to_dtype_("float32") \
             or dist.dtype == convert_np_dtype_to_dtype_("float64")
 
+    @property
+    def dim(self):
+        assert len(self.dist.shape) == 2
+        return self.dist.shape[1]
+
     def __call__(self):
         return self.dist
 
-    def loglikelihood(self, action):
-        assert False, "You cannot compute likelihood for a deterministic action!"
-
 
-def q_categorical_distribution(q_value, exploration_rate=0.0):
+def q_categorical_distribution(q_value):
     """
     Generate a PolicyDistribution object given a Q value.
-    We first construct a one-hot distribution according to the Q value,
-    and then add an exploration rate to get a probability.
+    We construct a one-hot distribution according to the Q value.
```

Comment:
How do you confirm that q_value is a policy distribution? 
Is it necessary to add a softmax layer here?

Reply:
q_value is not a distribution. It is a vector of Q values. This function tries to convert this vector to a one-hot distribution by taking max().

Reply:
https://github.com/PaddlePaddle/Paddle/blob/637827a5bc80d6e0a17466e44b087f91601539cb/python/paddle/fluid/layers/nn.py#L971
if it's a one-hot distribution, how could it be used to compute cross_entropy here ? 
https://github.com/yu239/RL/blob/ccadf0db160aa14018212b4bef0748d81c36093f/parl/framework/policy_distribution.py#L88

Reply:
If you look into the SimpleQ code, you will find that this one-hot distribution is only used for prediction but not for learning (at least in the Q-learning algorithm). The reason why we convert Q values to one-hot distribution is to make the prediction procedure the same for both AC and Q: namely, we derive a distribution from Q values and do sampling. The learning code of SimpleQ still uses Bellman Equation.